### PR TITLE
Property cache cleanups

### PR DESCRIPTION
**Pull request by havocp**
There's one bugfix (using the wrong cache from base classes), one optimization (always cache the merged __properties__ from base classes), one new warning (when a subclass tries to override a base class property, which seems to be a bug in the two cases TODO-commented here), and a refactor to decrease LOC by sharing cache logic. Also, rename accumulate_from_subclasses to be accurate.

### CODE REVIEW

File: bokeh/models/widgets/icons.py [hunk 1]
```diff
@@ -20,6 +20,7 @@ class Icon(AbstractIcon):
 
     """
 
+    # TODO this conflicts with PlotObject.name and should be renamed
```
Comment (mattpap) [inline bokeh/models/widgets/icons.py]: Perhaps `PlotObject.name` is too generic. Is it even used anywhere?

**Comment by mattpap** [comment]
Please add tests.

File: bokeh/models/widgets/icons.py (hunk 1 above)
Comment (havocp) [inline bokeh/models/widgets/icons.py]: See https://github.com/bokeh/bokeh/pull/3112

**Comment by havocp** [comment]
Tests added. I checked that the previous code fails, and it didn't fail exactly how I thought; there are two layers of bug. The first layer is that the caching didn't work at all, because of this Python surprise I don't understand yet:
```
class A(object):
    @classmethod
    def bar(cls):
        cls.__bar = 42
        # this assertion fails
        assert hasattr(cls, '__bar')

A.bar() # this fails, but comment it out and the below code passes.
A.__baz = 57
assert hasattr(A, '__baz')
```

I think it has something to do with mangling. Anyhow, since `cls.__bar` doesn't set a `__bar` that `hasattr(cls, '__bar')` can see, it doesn't matter that `hasattr` looks at base classes. But once the cache is made to work, it's necessary to use `__dict__` instead of `hasattr` as in my patch to avoid getting the cache from a base class.

The tests check for both problems.

**Comment by ahmadia** [comment]
It's a mangling issue:

```
In [8]: class A(object):
    @classmethod
    def bar(cls):
        cls.__bar = 42
        assert hasattr(cls, '_A__bar')
```

```
In [9]: x = A
In [10]: A.bar()
```

**Comment by ahmadia** [comment]
It looks like CPython mangles `__` access from within class method function calls, but not if you have an external reference to the object.  What's even more messed up is it seems to detect aliasing to the class object in method calls

```
class A(object):
    @classmethod
    def bar(cls, cls2):
        cls.__bar = 42
        cls2.__bar = 42
        print(hasattr(cls, '__bar'))
        print(hasattr(cls2, '__bar'))
```

```
A.bar(A)
False
False
```

**Comment by havocp** [comment]
According to http://stackoverflow.com/questions/13498151/python-double-underscore-mangling it's that the mangling happens at a textual/preprocessor level, so we mangle any identifier within the class definition. This is true even if the identifier is used to set a field on a _different_ class:
```
class C(object):
    pass

class A(object):
    @classmethod
    def bar(cls):
        cls.__bar = 42
        C.__bar = 57
        assert hasattr(cls, '_A__bar')
        assert hasattr(C, '_A__bar')

        setattr(cls, '__foo', 42)
        assert hasattr(cls, '__foo')
```
Strings aren't identifiers, so mangling doesn't affect `setattr` or `hasattr`.
The missing piece of information for me was that mangling is "compile time" (textual) rather than a runtime transformation.

File: bokeh/models/widgets/icons.py (hunk 1 above)
Comment (bryevdv) [inline bokeh/models/widgets/icons.py]: `name` is definitely used and I think we are stuck with it (looking back, maybe `alias` or `tag` would have been better). Regarding this, 'icon_name` or `type` or `kind` seem OK. Or even `icon`

**Comment by bryevdv** [comment]
This LGTM, thanks for cleaning this up @havocp

**Pull request merged**